### ISSUE TITLE
[BUG] z.discriminatedUnion should satisfy LiteralZodEventSchema

### ISSUE THREAD

**Issue by beefancohen** [inngest/inngest-js#396]
**Describe the bug**
Theres a type error when the data type in an event schema is a zod discriminated union.  We want to have different `metadata` types based on an enum type.

**To Reproduce**
Define an event schema with the data's type as `z.discriminatedUnion`
There should be a type error

**Expected behavior**
The types should pass.

**Code snippets / Logs / Screenshots**
```
const schema = z.discriminatedUnion("type", [
  z
    .object({
      type: z.literal(NotificationType.ProviderOpsTagged),
      metadata: z.object({ patientName: z.string(), patientId: z.number() }),
    })
    .merge(dataSchema),
  ...nonMetadataTypeSchemas,
]);

export const sendNotification = z.object({
  name: z.literal("notifications/send"),
  data: schema,
}) as unknown as LiteralZodEventSchema;
```

**Comment by spastorelli** [comment]
Is there currently a way around this?

Hitting the same issue with a project our team is currently working on. The same seems to happen with both zod unions and discriminated unions. Looks like the shim for the Zod types in the SDK doesn't handle them. This is the error I'm getting:
```
Property 'shape' is missing in type 'ZodDiscriminatedUnion<"type", [ZodObject<{ type: ZodLiteral<"import">; payload: ZodObject<{}, "strip", ZodTypeAny, {}, {}>; }, "strip", ZodTypeAny, { ...; }, { ...; }>, ZodObject<...>]>' but required in type 'AnyZodObject'.ts(1360)
zod.d.ts(25, 9): 'shape' is declared here.
```

**Comment by jpwilliams** [comment]
Hi @beefancohen and @spastorelli 👋

Fixing this in #432, which will add support for using `z.discriminatedUnion()` and `z.union()` when defining schemas. 🙂

**Comment by jpwilliams** [comment]
@beefancohen @spastorelli Released in [`inngest@3.7.3`](https://github.com/inngest/inngest-js/pull/433) 🙂
